File: 1036-rotting-oranges/1036-rotting-oranges.cpp
class Solution {
public:

    vector<pair<int,int>> movement = {{0,1},{0,-1},{1,0},{-1,0}};

    bool valid(int x,int y,int n,int m){
        return x>=0 and y>=0 and x<n and y<m;
    }

    int orangesRotting(vector<vector<int>>& grid) {
        int n=grid.size(), m=grid[0].size();
        queue<pair<int,pair<int,int>>> q;
        int ans=0;
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if(grid[i][j]==2) q.push({0,{i,j}});
            }
        }
        while(!q.empty()){
            int size=q.size();
            while(size--){
                int curTime = q.front().first;
                int x=q.front().second.first,y=q.front().second.second;
                q.pop();
                ans=max(ans,curTime);
                for(auto i:movement){
                    int xd = x+i.first,yd=y+i.second; 
                    if(valid(xd,yd,n,m) && grid[xd][yd]==1){
                        q.push({curTime+1,{xd,yd}});
                        grid[xd][yd]=2;
                    }   
                }
            }
        }
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if(grid[i][j]==1) return -1;
            }
        }
        return ans;
    }
};

File: 1036-rotting-oranges/1036-rotting-oranges.java
class Solution {

    boolean isValid(int i, int j, int n, int m){
        return i>=0 && j>=0 && i<n && j<m;
    }

    public int orangesRotting(int[][] grid) {
        int ans=0,n=grid.length,m=grid[0].length;
        Queue<ArrayList<Integer>> q = new LinkedList<>();
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if(grid[i][j]==2){
                    q.offer(new ArrayList<>(Arrays.asList(0,i,j)));
                }
            }
        }
        while(q.size()!=0){
            int size = q.size();
            while(size-->0){
                int time = q.peek().get(0);
                int x = q.peek().get(1);
                int y = q.peek().get(2);
                q.poll();
                ans=Math.max(ans,time);
                for(int i=-1;i<=1;i++){
                    for(int j=-1;j<=1;j++){
                        if(Math.abs(i)!=Math.abs(j)){
                            int xd = x+i, yd = y+j;
                            if(isValid(xd,yd,n,m)==true){
                                if(grid[xd][yd]==1){
                                    q.offer(new ArrayList<>(Arrays.asList(time+1,xd,yd)));
                                    grid[xd][yd]=2;
                                }
                            }
                        }
                    }
                }
            }
        }
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if(grid[i][j]==1){
                    return -1;
                }
            }
        }
        return ans;
    }
}